File: openhab/mounts/tmp/openhab/conf/automation/jsr223/python/personal/core_scenes.py
from __future__ import unicode_literals
from personal.core_helpers import reload_rules, get_date_string, get_date, get_location, sync_group_with_tags, has_same_location, METADATA_NAMESPACE, get_random_number, get_items_of_any_tags, create_helper_item, get_item_of_helper_item
from personal.core_scenes import SCENE_TAGS, get_scene_item_states, save_scene_item_states, trigger_scene_items, get_scene_states, SCENE_TRIGGER_TAGS
from core.triggers import when
from core.rules import rule
from core.jsr223.scope import ir, events, OFF, ON
from core.metadata import set_key_value, get_key_value, set_value
from org.openhab.core.types import UnDefType
from core.date import minutes_between, seconds_between, hours_between, ZonedDateTime
from org.openhab.core.model.script.actions import Log
from org.openhab.core.library.types import OnOffType


@rule("Core - Sync helper items of scenes", description="Core - Sync helper items", tags=['core', 'core-scenes'])
@when("Item added")
@when("Item updated")
@when("Item removed")
@when("System started")
@when("Item Core_Scenes_ReloadStates received update ON")
def sync_scenes_helpers(event):
    # Sync group gCore_Scenes_StateTriggers with scene trigger items
    sync_group_with_tags(
        ir.getItem("gCore_Scenes_StateTriggers"),
        SCENE_TRIGGER_TAGS
    )

    # Sync group gCore_Scenes with scene items
    sceneMembers = sync_group_with_tags(
        ir.getItem("gCore_Scenes"),
        SCENE_TAGS
    )

    # Check helper items
    for sceneMember in sceneMembers:
        # get scene location
        sceneLocation = get_location(sceneMember)

        # check metadata of context states
        contextStates = get_key_value(
            sceneMember.name,
            METADATA_NAMESPACE,
            'scenes',
            'context-states'
        )

        defaultContextStates = [
            ('reset', False)
        ]

        for context, state in defaultContextStates:
            if context not in contextStates:
                contextStates[context] = state

        set_key_value(
            sceneMember.name,
            METADATA_NAMESPACE,
            'scenes',
            'context-states',
            contextStates
        )

        # Create scene store trigger
        helper = create_helper_item(
            of=sceneMember,
            namespace='scenes',
            name='store-trigger',
            item_type='Number',
            category='settings',
            label="{0} überschreiben".format(
                sceneMember.label
            ),
            groups=['gCore_Scenes_StoreTriggers'],
            tags=['Point']
        )

        set_key_value(
            helper.name,
            'cellWidget',
            'label',
            '=items.{0}.title'.format(sceneMember.name)
        )

        set_key_value(
            helper.name,
            'cellWidget',
            'icon',
            'oh:settings'
        )

        set_key_value(
            helper.name,
            'cellWidget',
            'action',
            'options'
        )

        set_key_value(
            helper.name,
            'cellWidget',
            'actionItem',
            helper.name
        )

        set_key_value(
            helper.name,
            'listWidget',
            'action',
            'options'
        )

        set_key_value(
            helper.name,
            'listWidget',
            'actionItem',
            helper.name
        )

        set_key_value(
            helper.name,
            'listWidget',
            'icon',
            'oh:settings'
        )

        commandDescription = sceneMember.getCommandDescription()
        commandOptions = commandDescription.getCommandOptions() if commandDescription else []
        if commandOptions:
            set_key_value(
                helper.name,
                'stateDescription',
                'options',
                ','.join(
                    map(
                        lambda option: "{}={}".format(
                            option.getCommand(),
                            option.getLabel()
                        ),
                        commandOptions
                    )
                )
            )

            set_key_value(
                helper.name,
                'stateDescription',
                'pattern',
                '%d'
            )

        # Sync (Add) switches for each scene state
        stateValues = []
        for value, label in get_scene_states(sceneMember):
            stateValues.append(value)
            stateTriggerLabel = "{0}-Szene".format(label)
            stateTrigger = create_helper_item(
                of=sceneMember,
                namespace='scenes',
                name='trigger-state-{}'.format(value),
                item_type='Switch',
                category='party',
                label=stateTriggerLabel,
                groups=['gCore_Scenes_StateTriggers'],
                tags=SCENE_TRIGGER_TAGS
            )

            if stateTrigger.getLabel() != stateTriggerLabel:
                stateTrigger.setLabel(stateTriggerLabel)

            set_key_value(
                stateTrigger.name,
                METADATA_NAMESPACE,
                'scenes',
                'trigger-state',
                'to',
                value
            )

            set_key_value(
                stateTrigger.name,
                METADATA_NAMESPACE,
                'scenes',
                'trigger-state',
                'target-scene',
                sceneMember.name
            )

            set_key_value(
                stateTrigger.name,
                METADATA_NAMESPACE,
                'scenes',
                'trigger-state',
                'generated',
                True
            )

            set_value(
                stateTrigger.name,
                'ga',
                'Scene'
            )

            set_key_value(
                stateTrigger.name,
                'ga',
                'sceneReversible',
                False
            )

            if sceneLocation:
                set_key_value(
                    stateTrigger.name,
                    'ga',
                    'synonyms',
                    '{} in {}'.format(
                        stateTriggerLabel,
                        sceneLocation.label
                    )
                )

            set_key_value(
                stateTrigger.name,
                'cellWidget',
                'label',
                '=items.{0}.title'.format(stateTrigger.name)
            )

            set_key_value(
                stateTrigger.name,
                'cellWidget',
                'icon',
                'oh:party'
            )

            set_key_value(
                stateTrigger.name,
                'listWidget',
                'subtitle',
                '=items.{0}.displayState'.format(stateTrigger.name)
            )

            set_key_value(
                stateTrigger.name,
                'listWidget',
                'icon',
                'oh:party'
            )

        # Sync (Remove) switches for each scene state
        for stateTrigger in ir.getItem('gCore_Scenes_StateTriggers').members:
            triggerInfo = get_key_value(
                stateTrigger.name,
                METADATA_NAMESPACE,
                'scenes',
                'trigger-state'
            )

            # Do not remove manual created items that are just tagged wrong as it could be added manually to an existing (important) item.
            if not triggerInfo or 'generated' not in triggerInfo or not triggerInfo['generated']:
                continue

            if 'to' in triggerInfo and 'target-scene' in triggerInfo:
                try:
                    scene = ir.getItem(triggerInfo['target-scene'])
                    if scene and triggerInfo['to'] in map(
                        lambda (value, label): value,
                        get_scene_states(scene)
                    ):
                        continue
                except:
                    pass
            ir.remove(stateTrigger.name)

    # Reload rules
    reload_rules(
        activate_scene_triggers,
        activate_scene
    )
    reload_rules(
        store_scene_triggers,
        store_scene
    )
    reload_rules(
        manage_scenetriggers_triggers,
        manage_scenetriggers
    )


activate_scene_triggers = [
    "Member of gCore_Scenes received update"
]


@rule("Core - Activate scene.", description="Activate scene.", tags=['core', 'core-scenes', 'core-reload-activate_scene'])
@when("Member of gCore_Scenes received update")
def activate_scene(event):
    scene = ir.getItem(event.itemName)
    set_key_value(
        scene.name,
        METADATA_NAMESPACE,
        'scenes',
        "last-activation",
        get_date_string(ZonedDateTime.now())
    )
    trigger_scene_items(scene)


store_scene_triggers = [
    "Member of gCore_Scenes_StoreTriggers received update"
]


@rule("Core - Store scene.", description="Store scene.", tags=['core', 'core-scenes', 'core-reload-store_scene'])
@when("Member of gCore_Scenes_StoreTriggers received update")
def store_scene(event):
    sceneTrigger = ir.getItem(event.itemName)
    scene = get_item_of_helper_item(sceneTrigger)
    save_scene_item_states(scene, event.itemState.toFullString())


manage_scenetriggers_triggers = [
    "Member of gCore_Scenes_StateTriggers received update"
]


@rule("Core - Manage gCore_Scenes_StateTriggers to trigger scene.", description="Manage gCore_Scenes_StateTriggers to trigger scene.", tags=['core', 'core-scenes', 'core-reload-manage_scenetriggers'])
@when("Member of gCore_Scenes_StateTriggers received update")
def manage_scenetriggers(event):
    item = ir.getItem(event.itemName)
    triggerInfo = get_key_value(
        item.name,
        METADATA_NAMESPACE,
        'scenes',
        'trigger-state'
    )

    if (
        'states' in triggerInfo and
        item.state.toFullString() not in triggerInfo['states']
    ):
        return

    if 'to' in triggerInfo and 'target-scene' in triggerInfo:
        try:
            scene = ir.getItem(triggerInfo['target-scene'])
        except:
            return
        if (
            'from' in triggerInfo and
            not isinstance(scene.state, UnDefType) and
            triggerInfo['from'] != scene.state.toFullString()
        ):
            return

        lastActivation = get_key_value(
            scene.name,
            METADATA_NAMESPACE,
            'scenes',
            'last-activation'
        )

        if not lastActivation or (
            'hours-until-active' in triggerInfo and
            hours_between(
                get_date(lastActivation),
                ZonedDateTime.now()
            ) <= triggerInfo['hours-until-active']
        ) or (
            'minutes-until-active' in triggerInfo and
            minutes_between(
                get_date(lastActivation),
                ZonedDateTime.now()
            ) <= triggerInfo['minutes-until-active']
        ) or (
            'seconds-until-active' in triggerInfo and
            seconds_between(
                get_date(lastActivation),
                ZonedDateTime.now()
            ) <= triggerInfo['seconds-until-active']
        ):
            return

        events.postUpdate(scene, triggerInfo['to'])

File: openhab/mounts/tmp/openhab/conf/automation/js/core-scenes.js
const { rules, items, triggers, time } = require('openhab')
const { uniqBy, isEmpty } = require('lodash')
const {
  create_helper_item,
  get_all_semantic_items,
  get_items_of_any_tags,
  get_item_of_helper_item,
  DATETIME_FORMAT,
  stringifiedFloat,
  json_storage,
  sync_group_with_semantic_items,
  get_childs_with_condition,
  get_location,
  has_same_location
} = require(__dirname + '/core-helpers')
const { LIGHTS_EQUIPMENT_TAGS, LIGHTS_POINT_TAGS, is_on } = require(__dirname +
  '/core-lights')

const SCENE_TAGS = ['CoreScene']
const SCENE_TRIGGER_TAGS = ['CoreSceneTrigger']
const SceneTriggerStyle = {
  AUTO: 0,
  COMMAND_AND_UPDATE: 1,
  UPDATE: 2
}

function get_default_scene_state(scene) {
  const stateDescription = scene?.rawItem?.getStateDescription()?.getOptions()
  if (stateDescription?.length) {
    const command = stateDescription[0]?.getValue()
    if (command?.length > 0) {
      return command
    }
  }

  return undefined
}

function get_scene_states(scene) {
  const stateDescription = scene?.rawItem?.getStateDescription()?.getOptions()
  let obj = {}

  if (!stateDescription?.length) {
    return obj
  }

  for (const stateOption of stateDescription) {
    const command = stateOption.getValue()
    const label = stateOption.getLabel()
    if (command?.length > 0 && label?.length > 0) {
      obj[label] = command
    }
  }
  return obj
}

function get_scene_items(scene) {
  const sceneMembers = json_storage(scene)
    .get('scenes', 'custom-members')
    ?.map((i) => i.trim()) || ['default:true']

  const handleMember = (member) => {
    if (member.startsWith('tag:')) {
      return get_items_of_any_tags([member.substring(4)]).filter(
        (item) => item.type != 'GroupItem' && has_same_location(item, scene)
      )
    }

    // Use lights as default items in list.
    if (member.startsWith('default:')) {
      if (member == 'default:true') {
        return get_all_semantic_items(
          LIGHTS_EQUIPMENT_TAGS,
          LIGHTS_POINT_TAGS
        ).filter(
          (item) => item.type != 'GroupItem' && has_same_location(item, scene)
        )
      }
    }
    return get_childs_with_condition(member, (item) => item.type != 'GroupItem')
  }

  return uniqBy(
    sceneMembers.reduce(
      (memberList, newMember) => memberList.concat(handleMember(newMember)),
      []
    ),
    (item) => item.name
  )
}

function get_scene_item_states(scene) {
  let sceneState = Number.parseFloat(scene.state)
  sceneState = Number.isNaN(sceneState)
    ? get_default_scene_state(scene)
    : sceneState
  if (sceneState === undefined) {
    return []
  }

  const states = json_storage(scene).get(
    'scenes',
    'states',
    stringifiedFloat(sceneState)
  )

  return get_scene_items(scene).reduce((obj, newItem) => {
    if (states && states[newItem.name] !== undefined) {
      obj[newItem.name] = states[newItem.name]
    }
    return obj
  }, {})
}

function save_scene_item_states(scene, sceneState) {
  const sceneItems = get_scene_items(scene)
  sceneState = sceneState || get_default_scene_state(scene)

  if (sceneState !== undefined) {
    let sceneItemStates = {}
    for (const item of sceneItems) {
      sceneItemStates[item.name] = item.state
    }

    json_storage(scene).set(
      'scenes',
      'states',
      stringifiedFloat(sceneState),
      sceneItemStates
    )
  }
}

function trigger_scene_items(scene, style = SceneTriggerStyle.AUTO) {
  const itemStates = get_scene_item_states(scene)
  for (const itemName in itemStates) {
    try {
      const item = items.getItem(itemName)
      switch (style) {
        case SceneTriggerStyle.UPDATE:
          item.postUpdate(item.state)
          break
        case SceneTriggerStyle.COMMAND_AND_UPDATE:
          item.postUpdate(itemStates[itemName])
          item.sendCommand(itemStates[itemName])
          break
        default:
          item.postUpdate(itemStates[itemName])
          if (
            item.groupNames.includes('gCore_Lights_Switchables') &&
            is_on(item.state)
          ) {
            item.sendCommand(itemStates[itemName])
          }
      }
    } catch {
      console.log(
        'trigger_scene_items',
        `Could not set item state: Item ${item} does not exist.`
      )
    }
  }
}

function apply_context(context) {
  for (const scene of items.getItem('gCore_Scenes').members) {
    const contextState = json_storage(scene).get(
      'scenes',
      'context-states',
      context
    )

    if (scene.state == contextState) {
      break
    }

    const sceneStates = get_scene_states(scene)
    if (
      Object.values(sceneStates).some(
        (sceneState) => sceneState == contextState
      )
    ) {
      scene.postUpdate(stringifiedFloat(contextState))
    }
  }
}

function scriptLoaded() {
  rules.JSRule({
    name: 'sync_scenes_helpers',
    description: 'Core (JS) - Sync helper items of scenes',
    tags: ['core', 'core-scenes'],
    triggers: [
      triggers.GenericCronTrigger('30 0/5 * ? * * *'),
      triggers.SystemStartlevelTrigger(100),
      triggers.ItemStateUpdateTrigger('Core_Scenes_ReloadStates', 'ON')
    ],
    execute: (event) => {
      //Sync group gCore_Scenes_StateTriggers with scene trigger items
      sync_group_with_semantic_items(
        'gCore_Scenes_StateTriggers',
        SCENE_TRIGGER_TAGS
      )

      // Sync group gCore_Scenes with scene items
      const sceneMembers = sync_group_with_semantic_items(
        'gCore_Scenes',
        SCENE_TAGS
      )

      // Check helper items
      for (const sceneMember of sceneMembers) {
        // Create scene store trigger
        //TODO language
        create_helper_item(
          sceneMember,
          ['scenes', 'store-trigger'],
          'Number',
          'settings',
          `${sceneMember.label} überschreiben`,
          ['gCore_Scenes_StoreTriggers'],
          [],
          (helperItemName) => {
            let metadata = {}
            for (const path of ['listWidget', 'cellWidget']) {
              metadata[path] = {
                config: {
                  label: `=items.${sceneMember.name}.title`,
                  icon: 'f7:gear_alt',
                  action: 'options',
                  actionItem: helperItemName
                }
              }
            }

            const stateDescription = sceneMember.rawItem
              ?.getStateDescription()
              ?.getOptions()
            if (stateDescription?.length) {
              metadata.stateDescription = {
                config: {
                  options: (() => {
                    let optionArray = []
                    for (const option of stateDescription) {
                      optionArray.push(
                        `${option.getValue()}=${option.getLabel()}`
                      )
                    }
                    return optionArray.join(',')
                  })(),
                  pattern: '%d'
                }
              }
            }
          }
        )

        // Sync (Add) switches for each scene state
        const sceneLocation = get_location(sceneMember)
        const sceneStates = get_scene_states(sceneMember)
        for (const sceneName in sceneStates) {
          //TODO language
          const stateTriggerLabel = `${sceneName}-Szene`
          const stateTrigger = create_helper_item(
            sceneMember,
            ['scenes', `trigger-state-${sceneStates[sceneName]}`],
            'Switch',
            'party',
            stateTriggerLabel,
            ['gCore_Scenes_StateTriggers'],
            SCENE_TRIGGER_TAGS,
            (helperItemName) => {
              let metadata = {
                ga: {
                  value: 'Scene',
                  config: {
                    sceneReversible: false,
                    synonyms: sceneLocation
                      ? [
                          `${stateTriggerLabel} in ${sceneLocation.label}`,
                          sceneMember,
                          `${sceneMember}-Modus in ${sceneLocation.label}`
                        ].join(',')
                      : undefined
                  }
                }
              }

              for (const path of ['listWidget', 'cellWidget']) {
                metadata[path] = {
                  config: {
                    label: `=items.${helperItemName}.title`,
                    icon: 'f7:film',
                    subtitle: `=items.${helperItemName}.displayState`
                  }
                }
              }
              return metadata
            }
          )

          json_storage(stateTrigger).set('scenes', 'trigger-state', {
            to: sceneStates[sceneName],
            'target-scene': sceneMember.name,
            generated: true
          })
        }

        // Sync (Remove) switches for each scene state
        for (const stateTrigger of items.getItem('gCore_Scenes_StateTriggers')
          .members) {
          const triggerInfo = json_storage(stateTrigger).get(
            'scenes',
            'trigger-state'
          )

          // Do not remove manual created items that are just tagged wrong as it could be added manually to an existing (important) item.
          if (isEmpty(triggerInfo) || !triggerInfo['generated']) {
            continue
          }

          if (triggerInfo['to'] !== undefined && triggerInfo['target-scene']) {
            try {
              const scene = items.getItem(triggerInfo['target-scene'])
              if (
                scene &&
                Object.values(get_scene_states(scene)).some(
                  (sceneState) =>
                    sceneState == Number.parseFloat(triggerInfo['to'])
                )
              ) {
                continue
              }
            } catch {}
          }
          try {
            items.removeItem(stateTrigger.name)
          } catch {}
        }
      }
    }
  })

  rules.JSRule({
    name: 'activate_scene',
    description: 'Core (JS) - Activate scene.',
    tags: ['core', 'core-scenes'],
    triggers: [
      triggers.GroupStateUpdateTrigger('gCore_Scenes'),
      triggers.GroupCommandTrigger('gCore_Scenes')
    ],
    execute: (event) => {
      const scene = items.getItem(event.itemName)
      json_storage(scene).set(
        'scenes',
        'last-activation',
        time.ZonedDateTime.now().format(DATETIME_FORMAT)
      )
      trigger_scene_items(
        scene,
        event.triggerType == 'ItemStateUpdateTrigger'
          ? SceneTriggerStyle.AUTO
          : SceneTriggerStyle.COMMAND_AND_UPDATE
      )
    }
  })

  rules.JSRule({
    name: 'store_scene',
    description: 'Core (JS) - Store scene.',
    tags: ['core', 'core-scenes'],
    triggers: [triggers.GroupStateUpdateTrigger('gCore_Scenes_StoreTriggers')],
    execute: (event) => {
      const sceneTrigger = items.getItem(event.itemName)
      const scene = get_item_of_helper_item(sceneTrigger)
      save_scene_item_states(scene, sceneTrigger.state)
    }
  })

  rules.JSRule({
    name: 'manage_scenetriggers',
    description:
      'Core (JS) - Manage gCore_Scenes_StateTriggers to trigger scene.',
    tags: ['core', 'core-default_scene'],
    triggers: [triggers.GroupStateUpdateTrigger('gCore_Scenes_StateTriggers')],
    execute: (event) => {
      const item = items.getItem(event.itemName)
      const triggerInfo = json_storage(item).get('scenes', 'trigger-state')

      if (
        !triggerInfo ||
        !(triggerInfo['states'] || [item.state]).some(
          (state) => state == item.state
        )
      ) {
        return
      }

      if (triggerInfo['to'] && triggerInfo['target-scene']) {
        try {
          const scene = items.getItem(triggerInfo['target-scene'])

          if (
            triggerInfo['from'] &&
            triggerInfo['from'] != Number.parseFloat(scene.state)
          ) {
            return
          }

          const lastActivation = json_storage(scene).get(
            'scenes',
            'last-activation'
          )

          if (
            !lastActivation ||
            (triggerInfo['hours-until-active'] &&
              time.ZonedDateTime.parse(lastActivation, DATETIME_FORMAT).until(
                time.ZonedDateTime.now(),
                time.ChronoUnit.HOURS
              ) <= triggerInfo['hours-until-active']) ||
            (triggerInfo['minutes-until-active'] &&
              time.ZonedDateTime.parse(lastActivation, DATETIME_FORMAT).until(
                time.ZonedDateTime.now(),
                time.ChronoUnit.MINUTES
              ) <= triggerInfo['minutes-until-active']) ||
            (triggerInfo['seconds-until-active'] &&
              time.ZonedDateTime.parse(lastActivation, DATETIME_FORMAT).until(
                time.ZonedDateTime.now(),
                time.ChronoUnit.HOURS
              ) <= triggerInfo['seconds-until-active'])
          ) {
            return
          }

          scene.postUpdate(stringifiedFloat(triggerInfo['to']))
        } catch {}
      }
    }
  })
}

module.exports = {
  get_default_scene_state,
  get_scene_states,
  get_scene_items,
  get_scene_item_states,
  save_scene_item_states,
  trigger_scene_items,
  apply_context,
  SCENE_TAGS,
  SCENE_TRIGGER_TAGS,
  SceneTriggerStyle
}
